notebooks/Learning_Rate_Scheduler.ipynb

```diff
@@ -0,0 +1,3350 @@
+      "First 5 Learning Rates: [0.05, 0.05, 0.05, 0.05, 0.05]\n",
+      "\n",
+      "Scheduler: CosineAnnealingLR\n",
+      "Test Score: 0.4250\n",
+      "First 5 Learning Rates: [0.04999987663004646, 0.049999506521403426, 0.04999888967772375, 0.04999802610509541, 0.04999691581204153]\n"
+     ]
+    }
+   ],
+   "source": [
+    "class LRCaptureCallback(Callback):\n",
+    "    def on_epoch_end(self, net, **kwargs):\n",
+    "        # Log the learning rate of the optimizer\n",
+    "        lr = net.optimizer_.param_groups[0]['lr']\n",
+    "        if not hasattr(net.history, 'lr'):\n",
+    "            net.history.record('lr', lr)\n",
+    "        else:\n",
+    "            net.history[-1, 'lr'] = lr\n",
+    "\n",
+    "# Training function with learning rate tracking \n",
+    "def train_schedulers(X_train, X_test, y_train, y_test, lr=0.05, epochs=1000, hidden_dim=128):\n",
+    "    # Convert data to PyTorch tensors\n",
+    "    X_train = X_train.astype(np.float32)\n",
+    "    X_test = X_test.astype(np.float32)\n",
+    "    y_train = y_train.astype(np.float32).reshape(-1, 1)\n",
+    "    y_test = y_test.astype(np.float32).reshape(-1, 1)\n",
+    "\n",
+    "    # Split training data into training and validation sets\n",
+    "    X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.4, random_state=42)\n",
+    "\n",
+    "    # Define learning rate schedulers\n",
+    "    schedulers = [\n",
+    "        {\"name\": \"StepLR\", \"scheduler_class\": torch.optim.lr_scheduler.StepLR, \"params\": {\"step_size\": 100, \"gamma\": 0.3}},\n",
+    "        {\"name\": \"ReduceLROnPlateau\", \"scheduler_class\": torch.optim.lr_scheduler.ReduceLROnPlateau, \"params\": {\"mode\": \"max\", \"factor\": 0.7, \"patience\": 5}},\n",
+    "        {\"name\": \"CosineAnnealingLR\", \"scheduler_class\": torch.optim.lr_scheduler.CosineAnnealingLR, \"params\": {\"T_max\": 1000}},\n",
+    "    ]\n",
+    "\n",
+    "    results = {}\n",
+    "    for scheduler_info in schedulers:\n",
+    "        print(f\"\\nTraining with {scheduler_info['name']} scheduler...\")\n",
+    "\n",
+    "        # Set up the neural network with the specified scheduler\n",
+    "        net = NeuralNetClassifier(\n",
+    "            module=BreastCancerNet,\n",
+    "            max_epochs=epochs,\n",
+    "            lr=lr,\n",
+    "            optimizer=optim.SGD,\n",
+    "            criterion=nn.BCEWithLogitsLoss,\n",
+    "            callbacks=[\n",
+    "                LRScheduler(\n",
+    "                    policy=scheduler_info[\"scheduler_class\"],\n",
+    "                    **scheduler_info[\"params\"]\n",
+    "                ),\n",
+    "                LRCaptureCallback(),\n",
+    "            ],\n",
+    "            iterator_train__shuffle=True,\n",
+    "            train_split=None,\n",
+    "            module__input_dim=X_train.shape[1],\n",
+    "            module__hidden_dim=hidden_dim,\n",
```

Comment:
Personally, I would add `verbose=0` as an argument. That way, we don't have to scroll through multiple screens of the history log and can focus on the essential results. WDYT?

Reply:
I didn't know about this argument, thanks for suggesting!

Updated as discussed.